### ISSUE TITLE
Release 4.5.1 - Release Candidate 1 - Workload benchmarks metrics

### ISSUE THREAD

**Issue by fdalmaup** [wazuh/wazuh#18413]
The following issue aims to run all `workload benchmarks` for the current release candidate, report the results, and open new issues for any encountered errors.

## Workload benchmarks metrics information
|                                               |                                            |
|-----------------------------------------------|--------------------------------------------|
| **Main release candidate issue**              |                             #18366               |
| **Version**                                   |                          4.5.1                  |
| **Release candidate #**                       |                      RC1                      |
| **Tag**                                       |                         [v4.5.1-rc1](https://github.com/wazuh/wazuh/tree/v4.5.1-rc1)                  |
| **Previous Workload benchmarks metrics issue**|                       #18113                     |

## Test configuration
All tests will be run and workload performance metrics will be obtained for the following clustered environment configurations:

|                   |                    |
|-------------------|--------------------|
| **# Agents**      | **# Worker nodes** |
|         50000          |      25              |

## Test report procedure
All individual test checks must be marked as:
|                                  |                                            |
|---------------------------------|--------------------------------------------|
| Pass | The test ran successfully. |
| Xfail | The test was expected to fail and it failed. It must be properly justified and reported in an issue.  |
| Skip | The test was not run. It must be properly justified and reported in an issue.  |
| Fail | The test failed. A new issue must be opened to evaluate and address the problem. |

All test results must have one the following statuses: 
|                                  |                                            |
|---------------------------------|--------------------------------------------|
| :green_circle:  | All checks passed. |
| :red_circle:  | There is at least one failed check. |
| :yellow_circle:  | There is at least one expected fail or skipped test and no failures. |

Any failing test must be properly addressed with a new issue, detailing the error and the possible cause. It must be included in the `Fixes` section of the current release candidate main issue.

Any expected fail or skipped test must have an issue justifying the reason. All auditors must validate the justification for an expected fail or skipped test.

An extended report of the test results must be attached as a zip or txt. This report can be used by the auditors to dig deeper into any possible failures and details.

## Conclusions

All tests have been executed and the results can be found [here](https://github.com/wazuh/wazuh/issues/18413#issuecomment-1680591161).

All tests have passed and the fails have been reported or justified. I, therefore, conclude that this issue is finished and completed with failures for this release candidate.

## Auditors validation
The definition of done for this one is the validation of the conclusions and the test results from all auditors.

All checks from below must be accepted in order to close this issue.

- [x] @damarisg 
- [x] @fdalmaup

**Comment by fdalmaup** [comment]
# Issue Update

The tests were run and the following errors arose:
```
FAILED tests/performance/test_api/test_api_endpoints_performance.py::test_api_endpoints[put_/rootcheck]
FAILED tests/performance/test_api/test_api_endpoints_performance.py::test_api_endpoints[put_/syscheck]
FAILED tests/performance/test_api/test_api_endpoints_performance.py::test_api_endpoints[delete_/agents]
FAILED tests/performance/test_api/test_api_endpoints_performance.py::test_api_endpoints[delete_/security/users]
```

[Last run artifacts](https://github.com/wazuh/wazuh/files/12339580/artifacts283.zip)


We need to determine if this corresponds to an error related to the version being tested or the test run itself.

**Comment by fdalmaup** [comment]
# Issue Update

After several test runs, the same tests keep failing. This is related to several agents that cannot be connected by the Wazuh manager, therefore having unexpected results. 
The last executed test showed the following number of active agents in its `environment_health_check.txt` file:
```
Registered agents = 50000
Active agents = 30890
Never connected agents = 0
```
Checking the logs from the master node, the registration of the agents was verified so approximately 20000 go down afterward.

The number of workers corresponds to the ones set for the test.

**Comment by fdalmaup** [comment]
# Issue Update

After discussing with the team how these tests work and [the logs from the last run](https://github.com/wazuh/wazuh/files/12358896/last_run_artifacts.zip), several conclusions related to them have been reached. The main conclusion is the lack of resources in the performed builds.

## API performance results

---
|                                  |                                            |                                           |            
|--|--|--|
| **Endpoint test name**  | **Status** | **Issues Ref.** |
| GET /cluster/local/info | :green_circle: | |
| GET /cluster/nodes| :green_circle:  | |
| GET /cluster/healthcheck | :green_circle: | |
| GET /cluster/status | :green_circle: | |
| GET /cluster/local/config | :green_circle: | |
| GET /cluster/api/config | :green_circle: | |
| GET /cluster/configuration/validation | :green_circle: | |
| GET /manager/status | :green_circle: | |
| GET /manager/info | :green_circle: | |
| GET /manager/configuration | :green_circle: | |
| GET /manager/logs | :green_circle: | |
| GET /manager/api/config | :green_circle: | |
| GET /manager/configuration/validation | :green_circle: | |
| GET /mitre/groups | :green_circle: | |
| GET /mitre/metadata | :green_circle: | |
| GET /mitre/mitigations | :green_circle: | |
| GET /mitre/references | :green_circle: | |
| GET /mitre/software | :green_circle: | |
| GET /mitre/tactics | :green_circle: | |
| GET /mitre/techniques | :green_circle: | |
| GET /overview/agents | :green_circle:  | |
| GET /tasks/status | :green_circle: | |
| PUT /active-response | :yellow_circle: | https://github.com/wazuh/wazuh-qa/issues/1266 |
| PUT /rootcheck | :red_circle: | https://github.com/wazuh/wazuh/issues/16208 |
| PUT /syscheck | :red_circle: | https://github.com/wazuh/wazuh/issues/16208 |
| POST /groups | :green_circle: | |
| POST /security/users | :green_circle: | |
| POST /security/roles | :green_circle: | |
| POST /security/policies | :green_circle: | |
| POST /security/rules | :green_circle: | |
| PUT /agents/group | :yellow_circle: | Endpoint is not being tested (https://github.com/wazuh/wazuh-qa/issues/3665). This related issue was opened in the past: https://github.com/wazuh/wazuh/issues/13872 |
| PUT /agents/group/new_test_group/restart | :green_circle: | |
| PUT /agents/restart | :green_circle: |  |
| POST /agents | :green_circle: | |
| POST /agents/insert​|  :green_circle: | |
| POST /agents/insert​/quick | :green_circle: | |
| GET /agents | :green_circle: | |
| GET /agents/no_group | :green_circle: | |
| GET /agents/outdated | :green_circle: | |
| GET /decoders | :green_circle: | |
| GET /groups | :green_circle: | |
| GET /lists | :green_circle: | |
| GET /rules | :green_circle: | |
| GET /security/users | :green_circle: | |
| GET /security/roles | :green_circle: | |
| GET /security/policies | :green_circle: | |
| GET /security/rules | :green_circle: | |
| DELETE /groups | :green_circle: | |
| DELETE /agents | :red_circle: | _Failure explained below_ |
| DELETE /security/users | :red_circle: | _Failure explained below_ |
| DELETE /security/roles | :green_circle: | |
| DELETE /security/policies | :green_circle: | |
| DELETE /security/rules | :green_circle: | |
| PUT /manager/restart | :green_circle: | |
| PUT /cluster/restart | :green_circle: | |

### `PUT /rootcheck` and `PUT /syscheck` failures

The `environment_health_check` shows the number of active agents at a determined moment and not during the whole test, so this is not the most accurate way to determine the active agents:
```
[2023-08-15T15:08:09.776Z] INFO: ENVIRONMENT HEALTH CHECK REPORT
[2023-08-15T15:08:09.776Z] Registered agents = 50000
[2023-08-15T15:08:09.776Z] Active agents = 30890
[2023-08-15T15:08:09.776Z] Never connected agents = 0
```

The proper way of verifying the connected agents is with the logs, specifically the coordinator logs:
```
2023/08/15 15:06:59 DEBUG:    [Main]      Current backend stats: {'CLUSTER-Workload_benchmarks_metrics_B286_manager_24': 2000, 'CLUSTER-Workload_benchmarks_metrics_B286_manager_2': 2000, 'CLUSTER-Workload_benchmarks_metrics_B286_manager_12': 2000, 'CLUSTER-Workload_benchmarks_metrics_B286_manager_17': 2000, 'CLUSTER-Workload_benchmarks_metrics_B286_manager_18': 2000, 'CLUSTER-Workload_benchmarks_metrics_B286_manager_6': 2000, 'CLUSTER-Workload_benchmarks_metrics_B286_manager_16': 2000, 'CLUSTER-Workload_benchmarks_metrics_B286_manager_21': 2000, 'CLUSTER-Workload_benchmarks_metrics_B286_manager_23': 2000, 'CLUSTER-Workload_benchmarks_metrics_B286_manager_8': 2000, 'CLUSTER-Workload_benchmarks_metrics_B286_manager_22': 2000, 'CLUSTER-Workload_benchmarks_metrics_B286_manager_9': 2000, 'CLUSTER-Workload_benchmarks_metrics_B286_manager_13': 2000, 'CLUSTER-Workload_benchmarks_metrics_B286_manager_20': 2000, 'CLUSTER-Workload_benchmarks_metrics_B286_manager_11': 2000, 'CLUSTER-Workload_benchmarks_metrics_B286_manager_14': 2000, 'CLUSTER-Workload_benchmarks_metrics_B286_manager_7': 2000, 'CLUSTER-Workload_benchmarks_metrics_B286_manager_1': 2000, 'CLUSTER-Workload_benchmarks_metrics_B286_manager_15': 2000, 'CLUSTER-Workload_benchmarks_metrics_B286_manager_4': 2000, 'CLUSTER-Workload_benchmarks_metrics_B286_manager_3': 2000, 'CLUSTER-Workload_benchmarks_metrics_B286_manager_25': 2000, 'CLUSTER-Workload_benchmarks_metrics_B286_manager_5': 2000, 'CLUSTER-Workload_benchmarks_metrics_B286_manager_10': 2000, 'CLUSTER-Workload_benchmarks_metrics_B286_manager_19': 2000}
...
2023/08/15 15:08:01 DEBUG:    [Main]      Current backend stats: {'CLUSTER-Workload_benchmarks_metrics_B286_manager_24': 2000, 'CLUSTER-Workload_benchmarks_metrics_B286_manager_2': 2000, 'CLUSTER-Workload_benchmarks_metrics_B286_manager_12': 2000, 'CLUSTER-Workload_benchmarks_metrics_B286_manager_17': 2000, 'CLUSTER-Workload_benchmarks_metrics_B286_manager_18': 2000, 'CLUSTER-Workload_benchmarks_metrics_B286_manager_6': 2000, 'CLUSTER-Workload_benchmarks_metrics_B286_manager_16': 2000, 'CLUSTER-Workload_benchmarks_metrics_B286_manager_21': 2000, 'CLUSTER-Workload_benchmarks_metrics_B286_manager_23': 2000, 'CLUSTER-Workload_benchmarks_metrics_B286_manager_8': 2000, 'CLUSTER-Workload_benchmarks_metrics_B286_manager_22': 2000, 'CLUSTER-Workload_benchmarks_metrics_B286_manager_9': 2000, 'CLUSTER-Workload_benchmarks_metrics_B286_manager_13': 2000, 'CLUSTER-Workload_benchmarks_metrics_B286_manager_20': 2000, 'CLUSTER-Workload_benchmarks_metrics_B286_manager_11': 2000, 'CLUSTER-Workload_benchmarks_metrics_B286_manager_14': 2000, 'CLUSTER-Workload_benchmarks_metrics_B286_manager_7': 2000, 'CLUSTER-Workload_benchmarks_metrics_B286_manager_1': 2000, 'CLUSTER-Workload_benchmarks_metrics_B286_manager_15': 2000, 'CLUSTER-Workload_benchmarks_metrics_B286_manager_4': 2000, 'CLUSTER-Workload_benchmarks_metrics_B286_manager_3': 2000, 'CLUSTER-Workload_benchmarks_metrics_B286_manager_25': 2000, 'CLUSTER-Workload_benchmarks_metrics_B286_manager_5': 2000, 'CLUSTER-Workload_benchmarks_metrics_B286_manager_10': 2000, 'CLUSTER-Workload_benchmarks_metrics_B286_manager_19': 2000}
...
2023/08/15 15:09:03 DEBUG:    [Main]      Current backend stats: {'CLUSTER-Workload_benchmarks_metrics_B286_manager_24': 2000, 'CLUSTER-Workload_benchmarks_metrics_B286_manager_2': 2000, 'CLUSTER-Workload_benchmarks_metrics_B286_manager_12': 2000, 'CLUSTER-Workload_benchmarks_metrics_B286_manager_17': 2000, 'CLUSTER-Workload_benchmarks_metrics_B286_manager_18': 2000, 'CLUSTER-Workload_benchmarks_metrics_B286_manager_6': 2000, 'CLUSTER-Workload_benchmarks_metrics_B286_manager_16': 2000, 'CLUSTER-Workload_benchmarks_metrics_B286_manager_21': 2000, 'CLUSTER-Workload_benchmarks_metrics_B286_manager_23': 2000, 'CLUSTER-Workload_benchmarks_metrics_B286_manager_8': 2000, 'CLUSTER-Workload_benchmarks_metrics_B286_manager_22': 2000, 'CLUSTER-Workload_benchmarks_metrics_B286_manager_9': 2000, 'CLUSTER-Workload_benchmarks_metrics_B286_manager_13': 2000, 'CLUSTER-Workload_benchmarks_metrics_B286_manager_20': 2000, 'CLUSTER-Workload_benchmarks_metrics_B286_manager_11': 2000, 'CLUSTER-Workload_benchmarks_metrics_B286_manager_14': 2000, 'CLUSTER-Workload_benchmarks_metrics_B286_manager_7': 2000, 'CLUSTER-Workload_benchmarks_metrics_B286_manager_1': 2000, 'CLUSTER-Workload_benchmarks_metrics_B286_manager_15': 2000, 'CLUSTER-Workload_benchmarks_metrics_B286_manager_4': 2000, 'CLUSTER-Workload_benchmarks_metrics_B286_manager_3': 2000, 'CLUSTER-Workload_benchmarks_metrics_B286_manager_25': 2000, 'CLUSTER-Workload_benchmarks_metrics_B286_manager_5': 2000, 'CLUSTER-Workload_benchmarks_metrics_B286_manager_10': 2000, 'CLUSTER-Workload_benchmarks_metrics_B286_manager_19': 2000}
```
So, the total number of agents (50000) were connected to the different managers in a balanced way before and after the health check file generation.

Another valuable piece of information that can be extracted from the `cluster.log` files is related to the received `agent-info` chunks. At the moment of the health check file generation, the following was registered in one of the workers:
```
2023/08/15 15:08:03 INFO: [Worker CLUSTER-Workload_benchmarks_metrics_B286_manager_1] [Agent-info sync] Finished in 0.369s. Updated 17 chunks.
...
2023/08/15 15:08:13 INFO: [Worker CLUSTER-Workload_benchmarks_metrics_B286_manager_1] [Agent-info sync] Finished in 0.102s. Updated 3 chunks.
```
The `Agent-info sync` task sends the agent information (registered in `global.db`) from the workers to the master node in 65KB chunks. We compared the obtained values from the last RC tests and we found out that an expected number of chunks should be in the order of 30 chunks. It is probable that the `remoted` service lacked the necessary resources in the majority of managers and it was not able to connect with every agent until the other tasks ended. 

This being said, having registered agents but not active at the moment of the tests triggered the failures in the `PUT /rootcheck` and `PUT /syscheck` endpoints tests. These endpoints show as `failed_items` in their responses the agents that were not active and that could not perform the scan.  This behavior is fixed in #16208 and is part of the `4.6.0` release.

# `DELETE /agents` and `DELETE /security/users` failures

Both failures were due to timeouts when performing the tests.
<details><summary>Failure logs</summary>

    ______________________ test_api_endpoints[delete_/agents] ______________________
    
    test_case = {'body': {}, 'endpoint': '/agents', 'method': 'delete', 'parameters': {'agents_list': 'all', 'older_than': '0s', 'status': 'all'}}
    set_api_test_environment = None, api_healthcheck = None
    
        @pytest.mark.parametrize('test_case', test_data['test_cases'], ids=case_ids)
        def test_api_endpoints(test_case, set_api_test_environment, api_healthcheck):
            """Make an API request for each `test_case`.
        
            Args:
                test_case (dict): Dictionary with the endpoint to be tested and the necessary parameters for the test.
                set_api_test_environment (fixture): Fixture that modifies the API security options.
                api_healthcheck (fixture): Fixture used to check that the API is ready to respond requests.
            """
            # Apply xfails
            if test_case['endpoint'] in xfailed_items.keys() and \
                    test_case['method'] == xfailed_items[test_case['endpoint']]['method']:
                pytest.xfail(xfailed_items[test_case['endpoint']]['message'])
        
            base_url = api_details['base_url']
            headers = api_details['auth_headers']
            response = None
            try:
                response = getattr(requests, test_case['method'])(f"{base_url}{test_case['endpoint']}", headers=headers,
                                                                  params=test_case['parameters'], json=test_case['body'],
                                                                  verify=False)
    >           assert response.status_code == 200
    E           assert 500 == 200
    E             +500
    E             -200
    
    tests/performance/test_api/test_api_endpoints_performance.py:47: AssertionError
    ----------------------------- Captured stdout call -----------------------------
    Request elapsed time: 10.122s
    
    Status code: 500
    
    Full response: 
    {
      "title": "Wazuh Internal Error",
      "detail": "Timeout executing API request",
      "dapi_errors": {
        "master": {
          "error": "Timeout executing API request",
          "logfile": "WAZUH_HOME/logs/api.log"
        }
      },
      "error": 3021
    }
    __________________ test_api_endpoints[delete_/security/users] __________________
    
    test_case = {'body': {}, 'endpoint': '/security/users', 'method': 'delete', 'parameters': {'user_ids': 'all'}}
    set_api_test_environment = None, api_healthcheck = None
    
        @pytest.mark.parametrize('test_case', test_data['test_cases'], ids=case_ids)
        def test_api_endpoints(test_case, set_api_test_environment, api_healthcheck):
            """Make an API request for each `test_case`.
        
            Args:
                test_case (dict): Dictionary with the endpoint to be tested and the necessary parameters for the test.
                set_api_test_environment (fixture): Fixture that modifies the API security options.
                api_healthcheck (fixture): Fixture used to check that the API is ready to respond requests.
            """
            # Apply xfails
            if test_case['endpoint'] in xfailed_items.keys() and \
                    test_case['method'] == xfailed_items[test_case['endpoint']]['method']:
                pytest.xfail(xfailed_items[test_case['endpoint']]['message'])
        
            base_url = api_details['base_url']
            headers = api_details['auth_headers']
            response = None
            try:
                response = getattr(requests, test_case['method'])(f"{base_url}{test_case['endpoint']}", headers=headers,
                                                                  params=test_case['parameters'], json=test_case['body'],
                                                                  verify=False)
    >           assert response.status_code == 200
    E           assert 500 == 200
    E             +500
    E             -200
    
    tests/performance/test_api/test_api_endpoints_performance.py:47: AssertionError
    ----------------------------- Captured stdout call -----------------------------
    Request elapsed time: 10.122s
    
    Status code: 500
    
    Full response: 
    {
      "title": "Wazuh Internal Error",
      "detail": "Timeout executing API request",
      "dapi_errors": {
        "master": {
          "error": "Timeout executing API request",
          "logfile": "WAZUH_HOME/logs/api.log"
        }
      },
      "error": 3021
    }

</details>

The triggered tasks when executing these endpoints are slow if we consider the 50k registered agents, especially with scarce resources. It is expected that the response was not ready before the 10 seconds timeout. 
```
2023/08/15 15:14:12 ERROR: Timeout executing API request
2023/08/15 15:14:12 INFO: wazuh 172.31.58.220 "DELETE /agents" with parameters {"agents_list": "all", "status": "all", "older_than": "0s"} and body {} done in 10.068s: 500
2023/08/15 15:14:17 INFO: wazuh 172.31.2.137 "GET /cluster/nodes" with parameters {} and body {} done in 0.041s: 200
2023/08/15 15:14:22 ERROR: Timeout executing API request
2023/08/15 15:14:22 INFO: wazuh 172.31.58.220 "DELETE /security/users" with parameters {"user_ids": "all"} and body {} done in 10.071s: 500
```

Nevertheless, the triggered processes keep executing in the background after returning the Timeout response:
- `api.log`
```
2023/08/15 15:14:12 ERROR: Timeout executing API request
2023/08/15 15:14:12 INFO: wazuh 172.31.58.220 "DELETE /agents" with parameters {"agents_list": "all", "status": "all", "older_than": "0s"} and body {} done in 10.068s: 500
2023/08/15 15:14:17 INFO: wazuh 172.31.2.137 "GET /cluster/nodes" with parameters {} and body {} done in 0.041s: 200
2023/08/15 15:14:22 ERROR: Timeout executing API request
2023/08/15 15:14:22 INFO: wazuh 172.31.58.220 "DELETE /security/users" with parameters {"user_ids": "all"} and body {} done in 10.071s: 500
2023/08/15 15:14:31 INFO: wazuh 172.31.58.220 "DELETE /security/roles" with parameters {"role_ids": "all"} and body {} done in 9.341s: 200
```

- `ossec.log`
```
2023/08/15 15:14:22 wazuh-authd: INFO: Agent '42069' (1-XYuon7zi9hpIvMcQ-debian10) deleted (requested locally)
...
2023/08/15 15:14:30 wazuh-authd: INFO: Agent '4539' (1-aT1jxO6FbY4yZVWw-debian10) deleted (requested locally)
2023/08/15 15:14:31 wazuh-db: ERROR: There was an error assigning the groups to agent '32051'
2023/08/15 15:14:31 wazuh-db: WARNING: The groups were empty right after the set for agent '32051'
```

**Comment by damarisg** [comment]
LGTM!
